nav2_behaviors/plugins/drive_on_heading.cpp

```diff
@@ -0,0 +1,171 @@
+    this->stopRobot();
+    RCLCPP_WARN(
+      this->logger_,
+      "Exceeded time allowance before reaching the DriveOnHeading goal - Exiting DriveOnHeading");
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(
+      current_pose, *this->tf_, this->global_frame_, this->robot_base_frame_,
+      this->transform_tolerance_))
+  {
+    RCLCPP_ERROR(this->logger_, "Current robot pose is not available.");
+    return Status::FAILED;
+  }
+
+  double diff_x = initial_pose_.pose.position.x - current_pose.pose.position.x;
+  double diff_y = initial_pose_.pose.position.y - current_pose.pose.position.y;
+  double distance = sqrt(diff_x * diff_x + diff_y * diff_y);
+
+  feedback_->distance_traveled = distance;
+  this->action_server_->publish_feedback(feedback_);
+
+  if (distance >= std::fabs(command_x_)) {
+    this->stopRobot();
+    return Status::SUCCEEDED;
+  }
+
+  // TODO(mhpanah): cmd_vel value should be passed as a parameter
+  auto cmd_vel = std::make_unique<geometry_msgs::msg::Twist>();
+  cmd_vel->linear.y = 0.0;
+  cmd_vel->angular.z = 0.0;
+  cmd_vel->linear.x = command_speed_;
+
+  geometry_msgs::msg::Pose2D pose2d;
+  pose2d.x = current_pose.pose.position.x;
+  pose2d.y = current_pose.pose.position.y;
+  pose2d.theta = tf2::getYaw(current_pose.pose.orientation);
+
+  if (!isCollisionFree(distance, cmd_vel.get(), pose2d)) {
+    this->stopRobot();
+    RCLCPP_WARN(this->logger_, "Collision Ahead - Exiting DriveOnHeading");
+    return Status::FAILED;
+  }
+
+  this->vel_pub_->publish(std::move(cmd_vel));
+
+  return Status::RUNNING;
+}
+
+template<typename ActionT>
+bool DriveOnHeading<ActionT>::isCollisionFree(
+  const double & distance,
+  geometry_msgs::msg::Twist * cmd_vel,
+  geometry_msgs::msg::Pose2D & pose2d)
+{
+  // Simulate ahead by simulate_ahead_time_ in this->cycle_frequency_ increments
+  int cycle_count = 0;
+  double sim_position_change;
+  const double diff_dist = abs(command_x_) - distance;
+  const int max_cycle_count = static_cast<int>(this->cycle_frequency_ * simulate_ahead_time_);
+  geometry_msgs::msg::Pose2D init_pose = pose2d;
+  bool fetch_data = true;
+
+  while (cycle_count < max_cycle_count) {
+    sim_position_change = cmd_vel->linear.x * (cycle_count / this->cycle_frequency_);
+    pose2d.x = init_pose.x + sim_position_change * cos(init_pose.theta);
+    pose2d.y = init_pose.y + sim_position_change * sin(init_pose.theta);
+    cycle_count++;
+
+    if (diff_dist - abs(sim_position_change) <= 0.) {
+      break;
+    }
+
+    if (!this->collision_checker_->isCollisionFree(pose2d, fetch_data)) {
+      return false;
+    }
+    fetch_data = false;
+  }
+  return true;
+}
+
+}  // namespace nav2_behaviors
+
+#include "pluginlib/class_list_macros.hpp"
+PLUGINLIB_EXPORT_CLASS(nav2_behaviors::DriveOnHeading<nav2_behaviors::DriveOnHeadingAction>, nav2_core::Behavior)
```

Comment:
Use the msg definition here the `nav2_behaviors` prefix is really confusing, that works because you `using`'ed the message type in the namespace, but that makes it seem like there's a separate action class in the package. That would take me an hour as a normal reader to finally decrypt.